junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestClassInstanceConstructionParticipatingExtension.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2015-2024 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.api.extension;
+
+import static org.apiguardian.api.API.Status.DEPRECATED;
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+
+import org.apiguardian.api.API;
+import org.junit.jupiter.api.TestInstance;
+
+/**
+ * Interface for {@link Extension Extensions} that participate in the
+ * construction of test class instances.
+ *
+ * @since 5.12
+ * @see InvocationInterceptor#interceptTestClassConstructor
+ * @see ParameterResolver
+ * @see TestInstancePreConstructCallback
+ * @see TestInstancePostProcessor
+ * @see TestInstanceFactory
+ */
+@API(status = EXPERIMENTAL, since = "5.12")
+public interface TestClassInstanceConstructionParticipatingExtension extends Extension {
```

Comment:
:speech_balloon: This enforces that only potentially affected extensions can override the method.

Reply:
Other naming proposals:
- `TestInstantiationAwareExtension`
- `TestConstructionAwareExtension`